### PR TITLE
Email Verification & Password Reset

### PR DESCRIPTION
**Pull request by DOOduneye**

### CODE REVIEW

File: backend/src/transactions/auth.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+func GetOTP(db *gorm.DB, userID uuid.UUID) (*models.Verification, *errors.Error) {
+	var otp models.Verification
+	if err := db.Where("user_id = ? AND type = ?", userID, models.EmailVerificationType).First(&otp).Error; err != nil {
+		return nil, &errors.FailedToGetOTP
+	}
+
+	return &otp, nil
+}
+
+func DeleteOTP(db *gorm.DB, userID uuid.UUID) *errors.Error {
```
Comment (garrettladley) [inline backend/src/transactions/auth.go:36]: is this ever used? i couldn't find any usages.

what is the use case for this? (hinting that this likely shouldn't a blocking call at the end of a request but rather be executed by a background job)

File: backend/src/transactions/auth.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	if err := db.Create(&passwordReset).Error; err != nil {
+		return &errors.FailedToCreatePasswordReset
+	}
+
+	return nil
+}
+
+func DeletePasswordResetToken(db *gorm.DB, token string) *errors.Error {
```
Comment (garrettladley) [inline backend/src/transactions/auth.go:59]: same idea as `DeleteOTP`

**Review by garrettladley — CHANGES_REQUESTED** [review]
nitpicky future thinking and design comments, great work king @DOOduneye 👑 🐐

File: backend/src/auth/hash.go [hunk 3]
```diff
@@ -56,13 +79,13 @@ var (
...
+func CompareHash(data string, encodedHash string) (bool, error) {
 	p, salt, hash, err := decodeHash(encodedHash)
 	if err != nil {
 		return false, err
 	}
 
-	otherHash := argon2.IDKey([]byte(password), salt, p.iterations, p.memory, p.parallelism, p.keyLength)
+	otherHash := argon2.IDKey([]byte(data), salt, p.iterations, p.memory, p.parallelism, p.keyLength)
 
 	if subtle.ConstantTimeCompare(hash, otherHash) == 1 {
```
Comment (alderwhiteford) [inline backend/src/auth/hash.go:90]: recommend everyone to take a look at this, super interesting security vulnerability that is solved here

Comment (garrettladley) [reply backend/src/auth/hash.go:90]: I wrote this

File: backend/src/transactions/auth.go (hunk 1 above)
Comment (DOOduneye) [reply backend/src/transactions/auth.go:36]: It is used for deleting OTP, I think I may have forgot to use it.

It's extremely important to delete the OTP and password reset token from the db, this is stated by password reset resources (although the OTP is less important as it gets invalidated by use, the password reset token doesn't and also just have it unnecessarily laying around has no use removing it when used is much better

File: backend/src/transactions/auth.go (hunk 2 above)
Comment (DOOduneye) [reply backend/src/transactions/auth.go:59]: same as above

File: backend/src/services/auth.go [hunk 4]
```diff
@@ -83,30 +91,267 @@ func (a *AuthService) UpdatePassword(id string, passwordBody models.UpdatePasswo
 		return idErr
 	}
 
-	// TODO: Validate password
 	if err := a.Validate.Struct(passwordBody); err != nil {
 		return &errors.FailedToValidateUser
 	}
 
-	passwordHash, err := transactions.GetUserPasswordHash(a.DB, *idAsUint)
+	tx := a.DB.Begin()
```
[the remark this answers is not in the document]
Comment (DOOduneye) [reply backend/src/services/auth.go:98]: how else would you handle the set of transaction calls?

**Pull request merged**